barman/cloud_providers/aws_s3.py

```diff
@@ -462,18 +463,30 @@ class AwsCloudSnapshotInterface(CloudSnapshotInterface):
         https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-creating-snapshot.html
     """
 
-    def __init__(self, profile_name=None, region=None):
+    def __init__(self, profile_name=None, region=None, await_snapshots_timeout=3600):
         """
         Creates the client necessary for creating and managing snapshots.
 
         :param str profile_name: AWS auth profile identifier.
         :param str region: The AWS region in which snapshot resources are located.
+        :param int await_snapshots_timeout: The maximum time in seconds to wait for
+            snapshots to complete.
         """
         self.session = boto3.Session(profile_name=profile_name)
         # If a specific region was provided then this overrides any region which may be
         # defined in the profile
         self.region = region or self.session.region_name
         self.ec2_client = self.session.client("ec2", region_name=self.region)
+        self.await_snapshots_timeout = await_snapshots_timeout
+
+    def _get_waiter_config(self):
+        delay = 15
+        # Use ceil so that we always wait for at least the specified timeout
+        max_attempts = math.ceil(self.await_snapshots_timeout / delay)
```

Comment:
Should we treat a `await_snapshots_timeout` of `0` in a "special" way, or we just let it pass over?

Reply:
This is a good question.

If `await_snapshots_timeout` is `0` then currently `max_attempts` will work out as `0` as well.

As it happens, the current implementation will still make one attempt if we pass `"MaxAttempts": 0` in the `WaiterConfig`: https://github.com/boto/botocore/blob/547b20801770c8ea4255ee9c3b809fea6b9f6bc4/botocore/waiter.py#L379

However, despite being a de facto part of the boto3 interface, `WaiterConfig` [remains undocumented](https://github.com/boto/boto3/issues/712) so we shouldn't rely on the current implementation.

I think deliberately setting `MaxAttempts` to `1` in this case would be worthwhile just to be safe.

Reply:
Cool, agree with you!